modules/transport-netty4/src/main/java/org/elasticsearch/transport/CopyBytesSocketChannel.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright 2012 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.transport;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelOutboundBuffer;
+import io.netty.channel.RecvByteBufAllocator;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import org.elasticsearch.common.SuppressForbidden;
+
+import java.nio.ByteBuffer;
+import java.nio.channels.SocketChannel;
+
+import static io.netty.channel.internal.ChannelUtils.MAX_BYTES_PER_GATHERING_WRITE_ATTEMPTED_LOW_THRESHOLD;
+
+
+/**
+ * This class is adapted from {@link NioSocketChannel} class in the Netty project. It overrides the channel
+ * read/write behavior to ensure that the bytes are always copied to a thread-local direct bytes buffer. This
+ * happens BEFORE the call to the Java {@link SocketChannel} is issued.
```

Comment:
Maybe mention the motivation for the class in the java docs here and make it clear that it's motivated by not having Netty pool/create direct buffers and instead us managing just one? (IMO, otherwise this class and the logic around it are pretty hard to understand from the code).